stdlib/public/core/EmbeddedRuntime.swift

```diff
@@ -0,0 +1,121 @@
+  return r
+}
+
+@_cdecl("swift_slowAlloc")
+public func swift_slowAlloc(_ size: UInt, _ alignMask: UInt) -> UnsafeMutableRawPointer? {
+  let alignment: UInt
+  if alignMask == UInt.max {
+    alignment = UInt(_swift_MinAllocationAlignment)
+  } else {
+    alignment = alignMask + 1
+  }
+  return alignedAlloc(size: size, alignment: alignment)
+}
+
+@_cdecl("swift_slowDealloc")
+public func swift_slowDealloc(_ ptr: UnsafeMutableRawPointer?, _ size: UInt, _ alignMask: UInt) {
+  free(ptr)
+}
+
+@_silgen_name("swift_allocObject")
+public func swift_allocObject(metadata: UnsafeMutablePointer<ClassMetadata>, requiredSize: UInt, requiredAlignmentMask: UInt) -> UnsafeMutablePointer<HeapObject> {
+  let p = swift_slowAlloc(requiredSize, requiredAlignmentMask)!
+  let object = p.assumingMemoryBound(to: HeapObject.self)
+  object.pointee.metadata = metadata
+  object.pointee.refcount = 1
+  return object
+}
+
+@_silgen_name("swift_deallocClassInstance")
+public func swift_deallocClassInstance(object: UnsafeMutablePointer<HeapObject>, allocatedSize: UInt, allocatedAlignMask: UInt) {
+  free(object)
+}
+
+@_silgen_name("swift_initStackObject")
+public func swift_initStackObject(metadata: UnsafeMutablePointer<ClassMetadata>, object: UnsafeMutablePointer<HeapObject>) -> UnsafeMutablePointer<HeapObject> {
+  object.pointee.metadata = metadata
+  object.pointee.refcount = -1
+  return object
+}
+
+// TODO/FIXME: Refcounting and swift_once is not thread-safe, the following only works in single-threaded environments.
+
+@_silgen_name("swift_isUniquelyReferenced_nonNull_native")
+public func swift_isUniquelyReferenced_nonNull_native(object: UnsafeMutablePointer<HeapObject>) -> Bool {
+  // TODO/FIXME: Refcounting is not thread-safe, the following only works in single-threaded environments.
+  return object.pointee.refcount == 1
+}
+
+@_silgen_name("swift_retain")
+public func swift_retain(object: Builtin.RawPointer) -> Builtin.RawPointer {
+  if Int(Builtin.ptrtoint_Word(object)) == 0 { return object }
+  let o = UnsafeMutablePointer<HeapObject>(object)
+  // TODO/FIXME: Refcounting is not thread-safe, the following only works in single-threaded environments.
+  if o.pointee.refcount == -1 { return o._rawValue }
+  o.pointee.refcount += 1
+  return o._rawValue
+}
+
+@_silgen_name("swift_release")
+public func swift_release(object: Builtin.RawPointer) {
+  if Int(Builtin.ptrtoint_Word(object)) == 0 { return }
+  let o = UnsafeMutablePointer<HeapObject>(object)
+  // TODO/FIXME: Refcounting is not thread-safe, the following only works in single-threaded environments.
+  if o.pointee.refcount == -1 { return }
+  o.pointee.refcount -= 1
+  if o.pointee.refcount == 0 {
+    _swift_runtime_invoke_heap_object_destroy(o.pointee.metadata.pointee.destroy, o)
+  }
+}
+
+@_silgen_name("swift_beginAccess")
+public func swift_beginAccess(pointer: UnsafeMutableRawPointer, buffer: UnsafeMutableRawPointer, flags: UInt, pc: UnsafeMutableRawPointer) {
```

Comment:
At the end we need to provide some sort of exclusivity checking in the runtime, because this is a safety feature of swift. We could make that very simple, for example with a pre-allocated fixed-size access stack, e.g. of size 4.

Reply:
Added a TODO.